File: Backend/services/documentExtractor.js
const { PDFParse } = require('pdf-parse');
const mammoth = require('mammoth');
const textract = require('textract');
const { promisify } = require('util');
const XLSX = require('xlsx');
const Tesseract = require('tesseract.js');

const textractFromBuffer = promisify(textract.fromBufferWithName);

/**
 * Extract text from PDF file with format preservation
 * @param {Buffer} buffer - PDF file buffer
 * @returns {Promise<Object>} - Extracted text and metadata
 */
const extractFromPDF = async (buffer) => {
    try {
        // pdf-parse v2 API: instantiate PDFParse with data (buffer)
        const parser = new PDFParse({ data: buffer });

        // Get text content
        const result = await parser.getText();

        // Get page count (if available)
        let pageCount = 0;
        try {
            const pages = await parser.getPages();
            pageCount = pages.length;
        } catch (e) {
            // Page count not critical
        }

        return {
            text: result.text,
            metadata: {
                pages: pageCount,
                info: {}
            }
        };
    } catch (error) {
        throw new Error(`PDF extraction failed: ${error.message}`);
    }
};

/**
 * Extract text from DOCX file with format preservation
 * @param {Buffer} buffer - DOCX file buffer
 * @returns {Promise<Object>} - Extracted text and metadata
 */
const extractFromDOCX = async (buffer) => {
    try {
        const result = await mammoth.extractRawText({ buffer });

        // Mammoth also provides HTML conversion for better structure
        const htmlResult = await mammoth.convertToHtml({ buffer });

        return {
            text: result.value,
            html: htmlResult.value,
            metadata: {
                messages: result.messages
            }
        };
    } catch (error) {
        throw new Error(`DOCX extraction failed: ${error.message}`);
    }
};

/**
 * Extract text from DOC file (legacy format)
 * @param {Buffer} buffer - DOC file buffer
 * @param {String} filename - Original filename
 * @returns {Promise<Object>} - Extracted text
 */
const extractFromDOC = async (buffer, filename) => {
    try {
        const text = await textractFromBuffer(filename, buffer);

        return {
            text: text.trim(),
            metadata: {
                format: 'legacy DOC'
            }
        };
    } catch (error) {
        throw new Error(`DOC extraction failed: ${error.message}`);
    }
};

/**
 * Extract text from Excel file (XLSX, XLS)
 * @param {Buffer} buffer - Excel file buffer
 * @returns {Promise<Object>} - Extracted text and metadata
 */
const extractFromExcel = async (buffer) => {
    try {
        const workbook = XLSX.read(buffer, { type: 'buffer' });
        const sheetNames = workbook.SheetNames;
        let allText = [];
        let sheetData = [];

        // Extract text from all sheets
        for (const sheetName of sheetNames) {
            const worksheet = workbook.Sheets[sheetName];
            const jsonData = XLSX.utils.sheet_to_json(worksheet, { header: 1, defval: '' });
            
            // Convert sheet to text format
            const sheetText = jsonData.map(row => 
                row.filter(cell => cell !== '').join(' | ')
            ).join('\n');
            
            allText.push(`=== Sheet: ${sheetName} ===\n${sheetText}`);
            sheetData.push({
                name: sheetName,
                rows: jsonData.length,
                text: sheetText
            });
        }

        return {
            text: allText.join('\n\n'),
            metadata: {
                sheets: sheetNames,
                sheetCount: sheetNames.length,
                sheetData: sheetData
            }
        };
    } catch (error) {
        throw new Error(`Excel extraction failed: ${error.message}`);
    }
};

/**
 * Extract text from image using OCR
 * @param {Buffer} buffer - Image file buffer
 * @param {String} filename - Original filename
 * @returns {Promise<Object>} - Extracted text and metadata
 */
const extractFromImage = async (buffer, filename) => {
    try {
        console.log(`🔍 Starting OCR for image: ${filename}`);
        
        // Use Tesseract.js for OCR
        const { data: { text, words, paragraphs } } = await Tesseract.recognize(buffer, 'eng', {
            logger: m => {
                if (m.status === 'recognizing text') {
                    console.log(`   OCR Progress: ${Math.round(m.progress * 100)}%`);
                }
            }
        });

        console.log(`✅ OCR completed. Extracted ${text.length} characters`);

        return {
            text: text.trim(),
            metadata: {
                format: 'image (OCR)',
                wordCount: words?.length || 0,
                paragraphCount: paragraphs?.length || 0,
                confidence: words?.length > 0 ? 
                    (words.reduce((sum, w) => sum + (w.confidence || 0), 0) / words.length).toFixed(2) : 
                    'N/A'
            }
        };
    } catch (error) {
        throw new Error(`Image OCR extraction failed: ${error.message}`);
    }
};

/**
 * Main extraction function that routes to appropriate extractor
 * @param {Buffer} buffer - File buffer
 * @param {String} mimetype - File MIME type
 * @param {String} filename - Original filename
 * @returns {Promise<Object>} - Extracted text and metadata
 */
const extractText = async (buffer, mimetype, filename) => {
    let result;

    // PDF files
    if (mimetype === 'application/pdf') {
        result = await extractFromPDF(buffer);
    } 
    // Word DOCX files
    else if (mimetype === 'application/vnd.openxmlformats-officedocument.wordprocessingml.document') {
        result = await extractFromDOCX(buffer);
    } 
    // Word DOC files (legacy)
    else if (mimetype === 'application/msword') {
        result = await extractFromDOC(buffer, filename);
    }
    // Excel XLSX files
    else if (mimetype === 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet') {
        result = await extractFromExcel(buffer);
    }
    // Excel XLS files (legacy)
    else if (mimetype === 'application/vnd.ms-excel' || mimetype === 'application/excel') {
        result = await extractFromExcel(buffer);
    }
    // Image files (PNG, JPG, JPEG, etc.)
    else if (mimetype.startsWith('image/')) {
        result = await extractFromImage(buffer, filename);
    } 
    else {
        throw new Error(`Unsupported file type: ${mimetype}. Supported types: PDF, DOC, DOCX, XLS, XLSX, PNG, JPG, JPEG`);
    }

    // Clean up and format text
    const cleanedText = cleanText(result.text);

    // Estimate pages for non-PDF files (for page number references)
    let estimatedPages = result.metadata.pages || 1;
    if (!result.metadata.pages) {
        // Estimate: ~500 words per page
        const wordCount = countWords(cleanedText);
        estimatedPages = Math.max(1, Math.ceil(wordCount / 500));
        result.metadata.pages = estimatedPages;
    }

    return {
        text: cleanedText,
        originalText: result.text,
        html: result.html || null,
        metadata: result.metadata,
        wordCount: countWords(cleanedText)
    };
};

/**
 * Clean and normalize extracted text
 * @param {String} text - Raw extracted text
 * @returns {String} - Cleaned text
 */
const cleanText = (text) => {
    return text
        .replace(/\r\n/g, '\n')          // Normalize line endings
        .replace(/\n{3,}/g, '\n\n')      // Remove excessive blank lines
        .replace(/[ \t]+/g, ' ')         // Normalize spaces
        .trim();
};

/**
 * Count words in text
 * @param {String} text - Text to count
 * @returns {Number} - Word count
 */
const countWords = (text) => {
    return text.split(/\s+/).filter(word => word.length > 0).length;
};

module.exports = {
    extractText,
    extractFromPDF,
    extractFromDOCX,
    extractFromDOC,
    extractFromExcel,
    extractFromImage
};

File: Backend_py/services/document_extractor.py
import io
import logging
import os
import pdfplumber
import docx
import mammoth
import openpyxl
import pytesseract
from PIL import Image
from typing import Dict, Any, Optional
import re

logger = logging.getLogger(__name__)

# Allow explicit tesseract binary path via env (useful on Windows where PATH is often missing).
_tesseract_cmd = (os.getenv("TESSERACT_CMD") or "").strip()
if _tesseract_cmd:
    pytesseract.pytesseract.tesseract_cmd = _tesseract_cmd


def _ocr_pdf_page(page) -> str:
    """Best-effort OCR for PDF pages that have no extractable text."""
    try:
        # pdfplumber can rasterize a page to PIL image; OCR that image.
        pil_img = page.to_image(resolution=220).original
        text = pytesseract.image_to_string(pil_img)
        return text.strip() if text else ""
    except Exception as e:
        logger.warning("PDF page OCR fallback failed: %s", e)
        return ""

async def extract_text(buffer: bytes, mimetype: str, filename: str) -> Dict[str, Any]:
    result = {}
    
    if mimetype == 'application/pdf':
        result = await extract_from_pdf(buffer)
    elif mimetype == 'application/vnd.openxmlformats-officedocument.wordprocessingml.document':
        result = await extract_from_docx(buffer)
    elif mimetype == 'application/msword':
        # Legacy DOC is tricky in Python without external tools like antiword or libreoffice
        # For now, we'll try mammoth or a fallback
        result = await extract_from_docx(buffer) # Fallback attempt
    elif mimetype in ['application/vnd.openxmlformats-officedocument.spreadsheetml.sheet', 'application/vnd.ms-excel', 'application/excel']:
        result = await extract_from_excel(buffer)
    elif mimetype.startswith('image/'):
        result = await extract_from_image(buffer, filename)
    else:
        raise Exception(f"Unsupported file type: {mimetype}")

    cleaned_text = clean_text(result.get("text", ""))
    
    # Estimate pages
    estimated_pages = result.get("metadata", {}).get("pages", 1)
    if not estimated_pages or estimated_pages == 1:
        word_count = count_words(cleaned_text)
        estimated_pages = max(1, word_count // 500)
        if "metadata" not in result:
            result["metadata"] = {}
        result["metadata"]["pages"] = estimated_pages

    return {
        "text": cleaned_text,
        "originalText": result.get("text", ""),
        "html": result.get("html"),
        "metadata": result.get("metadata", {}),
        "wordCount": count_words(cleaned_text)
    }

async def extract_from_pdf(buffer: bytes) -> Dict[str, Any]:
    try:
        text = ""
        page_count = 0
        with pdfplumber.open(io.BytesIO(buffer)) as pdf:
            page_count = len(pdf.pages)
            for idx, page in enumerate(pdf.pages, start=1):
                page_text = page.extract_text()
                if page_text:
                    text += page_text + "\n"
                # Only extract tables that look like BOQ/product lists (have item/description/product/quantity)
                tables = page.extract_tables()
                page_had_table_rows = False
                if tables:
                    boq_header_keywords = ("item", "description", "product", "quantity", "qty", "rate", "amount", "unit", "specification", "make", "model")
                    for table in tables:
                        if not table:
                            continue
                        first_row = " ".join(str(c or "").lower() for c in table[0] if c).strip()
                        if not any(kw in first_row for kw in boq_header_keywords):
                            continue
                        for row in table:
                            if row and any(cell and str(cell).strip() for cell in row):
                                line = " | ".join(str(cell or "").strip() for cell in row)
                                if line.strip():
                                    text += line + "\n"
                                    page_had_table_rows = True
                        text += "\n"

                # OCR fallback for scanned/embedded-image PDF pages.
                # Trigger only when native extraction is poor to avoid noisy duplication.
                native_word_count = len(re.findall(r"\w+", page_text or ""))
                if native_word_count < 20 and not page_had_table_rows:
                    ocr_text = _ocr_pdf_page(page)
                    if ocr_text and len(re.findall(r"\w+", ocr_text)) >= 20:
                        logger.info("📄 OCR fallback used for PDF page %s", idx)
                        text += ocr_text + "\n"
        return {
            "text": text,
            "metadata": {"pages": page_count}
        }
    except Exception as e:
        logger.error(f"PDF extraction failed: {str(e)}")
        raise Exception(f"PDF extraction failed: {str(e)}")

async def extract_from_docx(buffer: bytes) -> Dict[str, Any]:
    try:
        # Get raw text using python-docx
        doc = docx.Document(io.BytesIO(buffer))
        text = "\n".join([para.text for para in doc.paragraphs])
        
        # Get HTML using mammoth
        html_result = mammoth.convert_to_html(io.BytesIO(buffer))
        
        return {
            "text": text,
            "html": html_result.value,
            "metadata": {"messages": [m.message for m in html_result.messages]}
        }
    except Exception as e:
        logger.error(f"DOCX extraction failed: {str(e)}")
        raise Exception(f"DOCX extraction failed: {str(e)}")

async def extract_from_excel(buffer: bytes) -> Dict[str, Any]:
    try:
        wb = openpyxl.load_workbook(io.BytesIO(buffer), data_only=True)
        all_text = []
        sheet_data = []
        
        for sheet_name in wb.sheetnames:
            ws = wb[sheet_name]
            rows = []
            for row in ws.iter_rows(values_only=True):
                filtered_row = [str(cell) for cell in row if cell is not None and str(cell).strip() != ""]
                if filtered_row:
                    rows.append(" | ".join(filtered_row))
            
            sheet_text = "\n".join(rows)
            all_text.append(f"=== Sheet: {sheet_name} ===\n{sheet_text}")
            sheet_data.append({
                "name": sheet_name,
                "rows": len(rows),
                "text": sheet_text
            })
            
        return {
            "text": "\n\n".join(all_text),
            "metadata": {
                "sheets": wb.sheetnames,
                "sheetCount": len(wb.sheetnames),
                "sheetData": sheet_data
            }
        }
    except Exception as e:
        logger.error(f"Excel extraction failed: {str(e)}")
        raise Exception(f"Excel extraction failed: {str(e)}")

async def extract_from_image(buffer: bytes, filename: str) -> Dict[str, Any]:
    try:
        img = Image.open(io.BytesIO(buffer))
        text = pytesseract.image_to_string(img)
        
        return {
            "text": text,
            "metadata": {
                "format": "image (OCR)",
                "filename": filename
            }
        }
    except Exception as e:
        logger.error(f"Image OCR failed: {str(e)}")
        raise Exception(f"Image OCR failed: {str(e)}")

def clean_text(text: str) -> str:
    if not text:
        return ""
    text = text.replace('\r\n', '\n')
    text = re.sub(r'\n{3,}', '\n\n', text)
    # text = re.sub(r'[ \t]+', ' ', text)  # DO NOT collapse multiple spaces as it destroys table structure
    return text.strip()

def count_words(text: str) -> int:
    if not text:
        return 0
    return len(re.findall(r'\w+', text))
